platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,233 @@
+        this.rosterState.put(rosterStateBuilder.build());
+
+        final Bytes currentCandidateRosterHash = currentRosterState.candidateRosterHash();
+        // remove existing candidate roster if present
+        this.rosterMap.remove(
+                ProtoBytes.newBuilder().value(currentCandidateRosterHash).build());
+
+        this.rosterMap.put(ProtoBytes.newBuilder().value(candidateRosterHash).build(), candidateRoster);
+        commit();
+    }
+
+    /**
+     * Determines the initial active roster based on the given software version and initial state.
+     * The active roster is obtained by adopting the candidate roster if a software upgrade is detected.
+     * Otherwise, the active roster is retrieved from the state or an exception is thrown if not found.
+     *
+     * @param version the software version of the current node
+     * @param initialState the initial state of the platform
+     * @return the active roster which will be used by the platform
+     */
+    @NonNull
+    @Override
+    public Roster determineActiveRoster(
+            @NonNull final SoftwareVersion version, @NonNull final ReservedSignedState initialState) {
+        final boolean softwareUpgrade = detectSoftwareUpgrade(version, initialState.get());
+
+        if (!softwareUpgrade) {
+            final Roster lastUsedActiveRoster = rosterStateAccessor.getActiveRoster();
+            // not in software upgrade mode (i.e., normal restart), return the
+            // active roster present in the state or throw if not found.
+            return Objects.requireNonNull(
+                    lastUsedActiveRoster, "Active Roster must be present in the state during normal network restart.");
+        }
+
+        // software upgrade is detected, adopt the candidate roster as the active roster.
+        final Roster activeRoster = adoptCandidateRoster(initialState.get().getRound() + 1);
+        commit();
+        return activeRoster;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getCandidateRoster() {
+        return rosterStateAccessor.getCandidateRoster();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Nullable
+    @Override
+    public Roster getActiveRoster() {
+        return rosterStateAccessor.getActiveRoster();
+    }
+
+    /**
+     * Adopts the candidate roster present in the state as the active roster.
+     *
+     * @param roundNumber the round number in which the candidate roster became active
+     */
+    @NonNull
+    private Roster adoptCandidateRoster(final long roundNumber) {
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final Roster candidateRoster = rosterMap.get(ProtoBytes.newBuilder()
+                .value(previousRosterState.candidateRosterHash())
+                .build());
+        if (candidateRoster == null) {
+            throw new IllegalStateException("Candidate roster not found in the state.");
```

Comment:
I'll try to clarify. There are multiple routes to arriving at an 'active' roster - candidateRoster adoption/existence of one in the state, GenesisRoster, OverrideRoster.

With inversion of control, the flow of arriving at this active roster will come from ServicesMain. One of such mechanisms is what's implemented in this API. When given a state, and called to determine an 'active' roster, will either 
adopt the candidate roster in the state if found on upgrade, or use a present active roster. The other two mechanisms are yet to be implemented.

As you noted earlier, there is not yet a reference to call this method, because the work lies in ServicesMain for the future (could be done by SpaceTime or whoever, doesn't matter as long as this API is there).

It is expected that a call to this method needs a candidate roster adopted on upgrade (of which it fails if not found), or simply restart with existing Roster in the state.